### DISCUSSION TITLE
cs50p: CS50's Introduction to Programming with Python - CS%E8%87%AA%E5%AD%A6%E6%8C%87%E5%8D%97

### DISCUSSION THREAD

**Discussion by giscus** [PKUFlyingPig/cs-self-learning#390] [Announcements]
# cs50p: CS50's Introduction to Programming with Python - CS%E8%87%AA%E5%AD%A6%E6%8C%87%E5%8D%97

CS自学指南

https://csdiy.wiki/%E7%BC%96%E7%A8%8B%E5%85%A5%E9%97%A8/CS50P/

**Comment by mancuoj** [comment]
准备三天（2/12~2/14）速通一下这个课，习题集的解答放在仓库里 https://github.com/mancuoj/CS50P

**Reply by mancuoj** [reply]
虽然但是，忙里偷闲，勉强算速通成功，说一下感想：难度相对来说比较低，比较适合 Python 入门或者有其他语言基础想学 Python 的同学，只看 Notes 然后做习题就可以了。

**Reply by silence143** [reply]
想问一下50和61A是二选一即可吗，现在研0，不知道学完这个能不能直接去看后面AI的课

**Reply by zivmax** [reply]
cs50p 属于 python 个语言的入门性质，cs61a虽然主要是讲解程序设计， python只是教学语言，但是有关 python 语言的知识和cs50p比起来可以算是高阶python编程。

**Reply by ShawnFan2** [reply]
想问一下，在python的学习过程中会遇到的问题就是当碰到一些问题时，不知道python中有没有这个函数，像week1中的bank.py，是怎么知道去用startswith这个函数的呢？假如是看官方文档的话，那python中那么多函数的功能全看一遍并记住感觉也并不现实

**Reply by mancuoj** [reply]
GPT 解千愁

**Reply by 1nram1** [reply]
您好，请问上过cs61A,上这门课还有意义吗

**Reply by Sakura-Flowing** [reply]
可以先听cs50py，再去听61a，反过来的话就不必要再听了，cs50也是相对基础的内容

**Comment by etherea1ity** [comment]
大佬week 5的test_plates.py的实现里少考虑了一种情况: beginning alphabetical checks。要再加一条断言如: assert is_valid("A1") == False才能通过所有样例似乎（我这里是这个情况）

**Reply by HaofanCao** [reply]
点了， 还好看到这条，beginning alphabetical checks试了好久都没过😭

**Comment by Chocolatieee0929** [comment]
有哪位大佬知道check50 出现的You seem to be missing these required files错误吗，相应的文件夹里明明就包含了对应的py文件，找了好久都没解决好T^T
```
$ check50 cs50/problems/2022/python/scourgify
Connecting......
You seem to be missing these required files:
scourgify.py
You are currently in: /workspaces/115781945, did you perhaps intend another directory?
$ ls cs50/problems/2022/python/scourgify
scourgify.py
```

**Reply by Chocolatieee0929** [reply]
我已经找到解决方式了，需要cd对应的文件夹里check，例如cd cs50/problems/2022/python/scourgify，再进行check50 cs50/problems/2022/python/scourgify

**Comment by Carefreehuang** [comment]
为什么我做week4的时候，import总是提示：Import "pyfiglet" could not be resolved，但我在终端也pip install pyfiglet==0.7了，并且pip list检查也能看到pyfiglet

**Reply by Andy-xiaokang** [reply]
不用管，直接运行就可以

**Reply by Mikeravel** [reply]
我也是，但直接运行没有出现问题

**Comment by teacup418** [comment]
有没有Java课程的推荐？

**Reply by taylover2016** [reply]
Java可以看看CS61B

**Comment by GuMengSama** [comment]
想问一下CS50P和CS61A是二选一即可么

**Reply by taylover2016** [reply]
CS50P更偏向于语言教学；CS61A是以Python为基础讲解计算机入门，应该是同CS50一个等级的。

**Comment by Andy-xiaokang** [comment]
the best one in cs50p [OOP](https://cs50.harvard.edu/python/2022/weeks/8/)

**Comment by taylover2016** [comment]
20230727更新，CS50要有一门专门的课程讲SQL了！泪目

**Reply by zhmingzhaung** [reply]
请问SQL的课程在哪儿

**Comment by Hornlessunicorns** [comment]
作业在哪里我想知道

**Reply by zivmax** [reply]
去官网上课，https://cs50.harvard.edu/python/2022/

**Comment by plafle** [comment]
在Week 4的little professor这道题中，我本地一直都是过的，然后交上去总是会出现”Did not find "9" in "Level: 6 + 6 =..."“这个错误。我看了一下我的actual output，感觉里面没有把我的输出显示全。
我的代码如下：
```python
import random


def main():
    l = get_level()
    cnt1 = 0
    cnt2 = 0
    lx = -1
    ly = -1
    p = 10
    i = 0
    while i<p:
        i += 1
        if (i == 13):
            print(f"Score: 9")
        x = generate_integer(l)
        y = generate_integer(l)
        if (lx >= 0):
            x = lx
            y = ly
        print(f"{x} + {y} = ")
        z = input("").strip()

        print(f"z = {x}")
        if (not z.isdigit() or int(z) != x + y):
            print("EEE")
            lx = x
            ly = y
            cnt2 += 1
            p += 1
            if (cnt2 == 3):
                print(f"{x} + {y} = {x + y}")
                lx = -1
                p -= 1
        else:
            cnt2 = 0
            cnt1 = cnt1 + 1

            lx = -1
    print(f"Score: 9")


def generate_integer(level):
    if level == 1:
        return random.randint(0,9)
    elif level == 2:
        return random.randint(10,99)
    else:
        return random.randint(100,999)

def get_level():
    while True:
        try:
            n = int(input("Level: ").strip())
            if (n <= 0 or n > 3):
                raise ValueError
            break
        except ValueError:
            pass
    return n



if __name__ == "__main__":
    main()
```

**Reply by plafle** [reply]
不好意思，上面这个没删掉调试信息，正确的版本是：
```python
import random


def main():
    l = get_level()
    cnt1 = 0
    cnt2 = 0
    lx = -1
    ly = -1
    p = 10
    i = 0
    while i<p:
        i += 1
        x = generate_integer(l)
        y = generate_integer(l)
        if (lx >= 0):
            x = lx
            y = ly
        print(f"{x} + {y} = ")
        z = input("").strip()

        if (not z.isdigit() or int(z) != x + y):
            print("EEE")
            lx = x
            ly = y
            cnt2 += 1
            p += 1
            if (cnt2 == 3):
                print(f"{x} + {y} = {x + y}")
                lx = -1
                p -= 1
        else:
            cnt2 = 0
            cnt1 = cnt1 + 1
    
            lx = -1
    print(f"Score: {cnt1}")


def generate_integer(level):
    if level == 1:
        return random.randint(0,9)
    elif level == 2:
        return random.randint(10,99)
    else:
        return random.randint(100,999)

def get_level():
    while True:
        try:
            n = int(input("Level: ").strip())
            if (n <= 0 or n > 3):
                raise ValueError
            break
        except ValueError:
            pass
    return n



if __name__ == "__main__":
    main()

**Comment by leishaozaici** [comment]
这是我的习题解答，欢迎友友参考https://github.com/leishaozaici/CS50P-2022

**Comment by april4xxxx** [comment]
把全部课程，笔记，习题，大作业做完了用了60h，供参考～

**Reply by Yuan-1028** [reply]
每次看这个课都想跪谢大佬啊！！！

**Reply by Armin1019** [reply]
想问下作业是在哪里做的呀，怎么检查是否正确

**Comment by hfeast** [comment]
讲的挺好的，在学习python的过程中多次强调自己去查看官方文档，培养英语能力

**Comment by Szjs-z** [comment]
https://learning.edx.org/course/course-v1:HarvardX+CS50+X/home

**Comment by YanXinyun-521** [comment]
大一刚入学小白想问下大家，https://github.com/mancuoj/CS50P这个网站打开每个文件夹里都是python的文件代码，请问有没有标准答案呢？有点看不懂这些题，该怎么写呢？

**Reply by mancuoj** [reply]
没有标准答案，只是参考答案，我不觉得这些题有什么会让人看不懂的地方，只能说没仔细看没仔细学

**Reply by YanXinyun-521** [reply]
请问在哪个网站上提交作业啊？怎么提交啊？

**Comment by YanXinyun-521** [comment]
搞了很久都没办法用check50检查作业的正确性，我用的是windows系统，打完代码想用网站里提到的“OJ"检查作业但是完全搞不出来这个系统
我刚开始自学，有没有人能教教我在哪个网站里可以测试我写的代码对不对啊？

**Reply by Raven061205** [reply]
作业里"How To Test" 一栏就有，在终端窗口直接输入给出的check50 cs50/problems/2022/python/xxx就可以

**Comment by xuehuixuehui** [comment]


**Reply by HaofanCao** [reply]
建议先去看cs50x

**Comment by lil-Journey** [comment]
自建日本节点服务器 月租5元不限流量 本人也在用 失效时可联系人工修复 qq:1165116287

**Comment by zlxFTH** [comment]
关于提交作业，课程提供的在线 vscode 网络不稳定，建议本地安装 check50 和 submit50 `pip install check50 submit50`，并确保 SSH 能连接到 github，这两工具的官方文档：<https://cs50.readthedocs.io/github/>
